apiserver/common.go

```diff
@@ -52,28 +65,33 @@ func GetSecrets(cluster *crv1.Pgcluster, ns string) ([]msgs.ShowUserSecret, erro
 	return output, err
 }
 
-func GetPodStatus(deployName, ns string) (string, string) {
+// GetReplicaPodStatus gets the status of all replica pods in the cluster. Specifically, using
+// the provided cluster name and namespace, it looks up all replica pod in the cluster, and then
+// provides a status for each pod ("Ready" or "Not Ready")
+func GetReplicaPodStatus(clusterName, ns string) (*ReplicaPodStatus, error) {
```

Comment:
@abrightwell I believe the reasoning is to be able to return a `nil` in an error case, so we don't have to pass the full struct around.

If there is something that is terribly breaking with this approach, we could refactor, but I suggest that we leave it and address it with future API updates.

Reply:
On it's own, I don't believe that is enough justification for keeping this a pointer. While it is not necessarily prescribed in the literature (though [this](https://github.com/golang/go/wiki/CodeReviewComments#pass-values) touches on the essence of the topic), the general guidance is not to worry about saving a few bytes as the cost of maintaining the pointer is far more expensive. Does it 'break' anything? No, but it's not necessarily good practice either.

Reply:
I was simply looking to return `nil` in the event that there was an error, and a `ReplicaPodStatus` was not created.  And while I could simply return an empty struct in an error scenario since the `error` itself should be the true indicator of an error condition, I still tend to like returning `nil` (e.g. for `ReplicaPodStatus`) to clearly indicate that no `ReplicaPodStatus` was created, and a pointer return value allows me to do so.

Also, the guidance is pretty vague around this area, i.e. it doesn't specifically say this is a bad practice or anything.  So I just read the docs as saying there is not really a reason to do so in most cases (i.e. no need to be paranoid about saving a few bytes), but if it does make sense to do so, there is no harm either (and therefore a viable solution).

Reply:
[https://golang.org/doc/effective_go.html#pointers_vs_values](https://golang.org/doc/effective_go.html#pointers_vs_values) as the arbiter...which reading through it, one can see it's open to some interpretation for the art of it.

But perhaps the reveal of this is that there is another status, i.e. `Unknown` when we are getting an error and as such `ReadyStatus` would have to remain a `string` :wink:

Reply:
> Also, the guidance is pretty vague around this area, i.e. it doesn't specifically say this is a bad practice or anything. So I just read the docs as saying there is not really a reason to do so in most cases (i.e. no need to be paranoid about saving a few bytes), but if it does make sense to do so, there is no harm either (and therefore a viable solution).

Right, I think we'd be hard pressed to find any official documentation that prescribes what to do in any case. The best we'll get is some 'guidance' as stated, specifically, that it doesn't makes sense in most cases. I personally feel like that this falls under the umbrella of most cases as I don't see anything special that justifies it, but that's just my opinion. Not something I feel so strongly about that I'd insist on holding up these changes.

Reply:
> But perhaps the reveal of this is that there is another status, i.e. Unknown when we are getting an error and as such ReadyStatus would have to remain a string

Or it's own type to ensure valid values. :-)

So, something like:

```
type readyStatus string

const (
  ReadyStatusReady readyStatus = "ready"
  ReadyStatusNotReady readyStatus = "not ready"
  ReadyStatusUnknown readyStatus = "unknown"
)

type ReplicaPodStatus struct {
...
  ReadyStatus readyStatus
}
```

But perhaps that's all a discussion for another day. :-)